File: rust/src/calibrate.rs
//! One arc's probe ladder, fitted (spec §2.3, §2.6).
//!
//! A port of `core/calibrate.py`, which stays the reference: it is the only
//! place a `B` is produced, and `B >= 0` plus `clamped => cap finite` are its
//! postconditions rather than something the solver guards against.
//!
//! The arrays here are six elements long.  That is the whole reason this is
//! worth compiling: the Python runs `np.diff`, `np.interp` and `np.concatenate`
//! over six floats, and a single `np.diff` pair costs 5 us against nanoseconds
//! of arithmetic -- 50 us per call, 736 calls a quote.  Nothing about the
//! arithmetic is hard; the dispatch was the cost.

pub const DRIFT_TOL: f64 = 0.25;
pub const SATURATION_TOL: f64 = 1e-9;
pub const DUPLICATE_TOL: f64 = 1e-6;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Flag {
    None,
    DividedDiff,
    Structural,
    Clamped,
    Both,
}

impl Flag {
    pub fn as_str(self) -> &'static str {
        match self {
            Flag::None => "NONE",
            Flag::DividedDiff => "DIVIDED_DIFF",
            Flag::Structural => "STRUCTURAL",
            Flag::Clamped => "CLAMPED",
            Flag::Both => "BOTH",
        }
    }
}

#[derive(Debug)]
pub struct Calibration {
    pub a: f64,
    pub b: f64,
    pub cap: f64,
    pub clamped: bool,
    pub convex_flag: bool,
    pub flag: Flag,
    pub drift: f64,
    pub eta: f64,
    pub split_hint: bool,
    pub calib_delta: f64,
    pub tangent_delta: f64,
    pub note: &'static str,
}

#[derive(Debug)]
pub struct CalibrationError(pub String);

/// `f[x_k, x_k+1, x_k+2]` for each consecutive triple.
fn second_divided(x: &[f64], y: &[f64], out: &mut Vec<f64>) {
    out.clear();
    if x.len() < 3 {
        return;
    }
    for k in 0..(x.len() - 2) {
        let first_a = (y[k + 1] - y[k]) / (x[k + 1] - x[k]);
        let first_b = (y[k + 2] - y[k + 1]) / (x[k + 2] - x[k + 1]);
        out.push((first_b - first_a) / (x[k + 2] - x[k]));
    }
}

/// `numpy.interp` for a single point on an increasing `x`.
///
/// Matches numpy's own form -- `slope * (xq - x[i]) + y[i]`, clamped to the
/// endpoints outside the range -- because the result feeds `B`, and the two
/// spellings of a linear interpolation do not agree in the last bit.
fn interp(xq: f64, x: &[f64], y: &[f64]) -> f64 {
    let n = x.len();
    if n == 0 {
        return f64::NAN;
    }
    if xq <= x[0] {
        return y[0];
    }
    if xq >= x[n - 1] {
        return y[n - 1];
    }
    let mut lo = 0usize;
    let mut hi = n - 1;
    while hi - lo > 1 {
        let mid = (lo + hi) / 2;
        if x[mid] <= xq {
            lo = mid;
        } else {
            hi = mid;
        }
    }
    let slope = (y[lo + 1] - y[lo]) / (x[lo + 1] - x[lo]);
    let res = slope * (xq - x[lo]) + y[lo];
    if res.is_finite() {
        res
    } else {
        slope * (xq - x[lo + 1]) + y[lo + 1]
    }
}

fn sign(v: f64) -> i32 {
    if v > 0.0 {
        1
    } else if v < 0.0 {
        -1
    } else {
        0
    }
}

#[allow(clippy::too_many_arguments)]
pub fn calibrate(
    deltas: &[f64],
    quotes: &[f64],
    delta_bar: Option<f64>,
    structural_flag: bool,
    drift_tol: f64,
    cap_in: Option<f64>,
    f_at_cap_in: Option<f64>,
    quantum: f64,
) -> Result<Calibration, CalibrationError> {
    if deltas.len() != quotes.len() {
        return Err(CalibrationError(
            "deltas and quotes must be 1-D and the same length".into(),
        ));
    }
    if deltas.len() < 2 {
        return Err(CalibrationError(format!(
            "need at least 2 probes, got {}",
            deltas.len()
        )));
    }
    for k in 1..deltas.len() {
        if !(deltas[k] > deltas[k - 1]) {
            return Err(CalibrationError("deltas must be strictly increasing".into()));
        }
    }
    for &v in quotes {
        if !(v > 0.0) {
            return Err(CalibrationError(
                "all quotes must be positive; drop failed probes first".into(),
            ));
        }
    }

    // Two nodes at the same size are one probe, and the wall test below cannot
    // read them as anything but saturation.
    let mut x: Vec<f64> = Vec::with_capacity(deltas.len());
    let mut y: Vec<f64> = Vec::with_capacity(quotes.len());
    x.push(deltas[0]);
    y.push(quotes[0]);
    for k in 1..deltas.len() {
        if deltas[k] - deltas[k - 1] > deltas[k - 1] * DUPLICATE_TOL {
            x.push(deltas[k]);
            y.push(quotes[k]);
        }
    }
    if x.len() < 2 {
        return Err(CalibrationError(
            "need at least 2 probes at distinct sizes, got 1 after collapsing duplicates".into(),
        ));
    }

    let tangent_delta = x[0];
    let mut a = y[0] / x[0];

    // --- capacity wall (§2.3 rule 2) -------------------------------------
    let mut saturated_at: Option<f64> = None;
    for k in 1..x.len() {
        if y[k] <= y[k - 1] * (1.0 + SATURATION_TOL) {
            saturated_at = Some(x[k - 1]);
            x.truncate(k);
            y.truncate(k);
            break;
        }
    }
    if let Some(wall) = saturated_at {
        if x.len() == 1 {
            return Ok(Calibration {
                a,
                b: 0.0,
                cap: wall,
                clamped: true,
                convex_flag: true,
                flag: Flag::Clamped,
                drift: 0.0,
                eta: f64::NAN,
                split_hint: false,
                calib_delta: wall,
                tangent_delta,
                note: "SATURATED",
            });
        }
    }

    let mut cap = match (saturated_at, cap_in) {
        (Some(wall), None) => Some(wall),
        (Some(wall), Some(c)) => Some(c.min(wall)),
        (None, c) => c,
    };
    let d_bar = delta_bar.unwrap_or_else(|| x[x.len() - 1]);
    let f_bar = interp(d_bar, &x, &y);
    let mut b = 2.0 * (a * d_bar - f_bar) / (d_bar * d_bar);

    // --- what the output token's own resolution can fake ------------------
    let mut noise = 0.0;
    if quantum > 0.0 {
        noise = 2.0 * (quantum / x[0]) / d_bar + 2.0 * quantum / (d_bar * d_bar);
    }
    let quantised = b < 0.0 && -b <= noise;
    if quantised {
        b = noise;
    }

    // Flag detection uses every probe, so a convex patch anywhere is caught.
    let mut xs = Vec::with_capacity(x.len() + 1);
    let mut ys = Vec::with_capacity(y.len() + 1);
    xs.push(0.0);
    ys.push(0.0);
    xs.extend_from_slice(&x);
    ys.extend_from_slice(&y);
    let mut dd = Vec::new();
    second_divided(&xs, &ys, &mut dd);
    let any_positive = dd.iter().any(|&v| v > 0.0);
    let mixed_signs = dd.len() > 1 && {
        let first = sign(dd[0]);
        dd.iter().any(|&v| sign(v) != first)
    };
    let numeric_flag = any_positive || mixed_signs;

    // DRIFT and eta use the local window {0, d/4, d/2, d} of §2.3.
    let mut drift = 0.0;
    let mut eta = f64::NAN;
    if x.len() >= 3 {
        let tail = x.len() - 3;
        let xs_local = [0.0, x[tail], x[tail + 1], x[tail + 2]];
        let ys_local = [0.0, y[tail], y[tail + 1], y[tail + 2]];
        let mut d_local = Vec::new();
        second_divided(&xs_local, &ys_local, &mut d_local);
        if d_local[0] != 0.0 {
            drift = d_local[1] / d_local[0] - 1.0;
        }
        if d_local[1] != 0.0 {
            let d3 = (d_local[1] - d_local[0]) / (xs_local[3] - xs_local[0]);
            eta = 3.0 * a * d3 / (2.0 * d_local[1] * d_local[1]);
        }
    }

    let mut note = if saturated_at.is_some() { "SATURATED" } else { "" };
    if quantised {
        note = "QUANTISED";
    }
    let clamped = b <= 0.0 || saturated_at.is_some();
    let mut flag = match (numeric_flag, structural_flag) {
        (true, true) => Flag::Both,
        (true, false) => Flag::DividedDiff,
        (false, true) => Flag::Structural,
        (false, false) => Flag::None,
    };

    let mut f_at_cap = f_at_cap_in;
    if clamped {
        // §2.3 zero-curvature clamp: the chord, not the tangent.
        if cap.map_or(true, |c| !c.is_finite()) {
            cap = Some(x[x.len() - 1]);
            f_at_cap = Some(y[y.len() - 1]);
            note = "CAP_FROM_LADDER";
        }
        let cap_value = cap.unwrap();
        let f_cap = match f_at_cap {
            Some(v) => v,
            None => interp(cap_value, &x, &y),
        };
        if !(f_cap > 0.0) {
            return Err(CalibrationError(
                "clamped arc needs a positive f(cap) for the chord".into(),
            ));
        }
        a = f_cap / cap_value;
        b = 0.0;
        flag = if flag == Flag::None { Flag::Clamped } else { Flag::Both };
    }

    if !(a > 0.0) {
        return Err(CalibrationError(format!("a must be positive, got {a}")));
    }
    Ok(Calibration {
        a,
        b: b.max(0.0),
        cap: cap.unwrap_or(f64::INFINITY),
        clamped,
        convex_flag: numeric_flag || structural_flag || clamped,
        flag,
        drift,
        eta,
        split_hint: drift.abs() > drift_tol,
        calib_delta: d_bar,
        tangent_delta,
        note,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn cpmm(x0: f64, y0: f64, sizes: &[f64]) -> (Vec<f64>, Vec<f64>) {
        let q = sizes.iter().map(|&d| y0 * d / (x0 + d)).collect();
        (sizes.to_vec(), q)
    }

    #[test]
    fn a_cpmm_ladder_fits_a_positive_curvature() {
        let (x, y) = cpmm(1e6, 1e6, &[1.0, 1e2, 1e3, 1e4]);
        let got = calibrate(&x, &y, None, false, DRIFT_TOL, None, None, 0.0).unwrap();
        assert!(got.a > 0.0 && got.b > 0.0, "{got:?}");
        assert!(!got.clamped);
    }

    #[test]
    fn a_wall_becomes_a_finite_cap() {
        let x = vec![1.0, 10.0, 100.0, 1000.0];
        let y = vec![1.0, 10.0, 11.472806, 11.472806];
        let got = calibrate(&x, &y, None, false, DRIFT_TOL, None, None, 0.0).unwrap();
        assert!(got.clamped, "{got:?}");
        assert!(got.cap.is_finite(), "a clamped arc needs a finite cap");
        assert_eq!(got.b, 0.0);
    }

    #[test]
    fn duplicate_sizes_collapse_rather_than_reading_as_a_wall() {
        let x = vec![1.0, 1.0 + 1e-12, 100.0];
        let y = vec![1.0, 1.0, 99.0];
        let got = calibrate(&x, &y, None, false, DRIFT_TOL, None, None, 0.0).unwrap();
        assert!(!got.clamped, "the duplicate read as saturation: {got:?}");
    }

    #[test]
    fn a_short_ladder_is_refused() {
        assert!(calibrate(&[1.0], &[1.0], None, false, DRIFT_TOL, None, None, 0.0).is_err());
    }
}

File: src/erouter/core/calibrate.py
"""Probe-ladder calibration and the non-concavity detector (spec §2.3, §2.6).

Four evaluations of the true `f` -- the free origin plus three quotes --
determine `(a, B)` *and* say whether the quadratic model is admissible at all:

    a = f(delta_eps) / delta_eps
    B = 2 (a d_bar - f(d_bar)) / d_bar^2          secant fit  (M2)

The secant fit matches the true curve exactly at 0 and at `d_bar`.  Use it, not
the tangent `|f''(0)|`: for CPMM the tangent is off by `-theta^2` (100 bp at
theta = 10%) while the secant overshoots by only `theta_bar^2/4` (2.5 bp).  The
tangent under-promises, which is the safe direction, but the secant is 4x
tighter and the quote is re-verified on-chain anyway.

**The ladder is a detector, not a model.**  The divided differences are
diagnostics; they are never substituted into the solve.  Extra probes do
correctly *see* dynamic-fee non-concavity, but no Taylor order is
simultaneously faithful to increasing returns and admissible in a convex
program: if the fit reports `f'' > 0` anywhere then `G = nu a / B < 0`, the
Laplacian stops being PSD, and the certificate is void.  So the element law
stays quadratic and `B` is clamped at zero, never negative.
"""

from __future__ import annotations

import math
import os
from dataclasses import dataclass

import numpy as np

from . import accel as _accel
from .types import FlagReason

DRIFT_TOL = 0.25
#: The compiled fit is opt-in on the same switch as the compiled solve, so
#: a session runs both ports or neither and a comparison means something.
_ACCEL_ON = os.environ.get("EROUTER_ACCEL", "") == "1"
# Two probes an order of magnitude apart returning the *same* output is not a
# curve, it is a wall: the venue has handed over everything it has.  Measured on
# a LLAMMA WETH market, `get_dy` returned the same 11.472806 crvUSD across a
# thousandfold range of inputs.  Fitting a rate through saturated points reads
# the marginal price sixfold low, and in §4's log-price fit that one arc drags
# the whole frame.
SATURATION_TOL = 1e-9
# How far apart two ladder nodes must be to count as two probes at all.
#
# The wall test asks whether a bigger probe bought more, which two probes at the
# *same* size cannot answer: they return the same quote, and "bought nothing
# more" then clamps a perfectly healthy arc.  `merge` already drops duplicates,
# but it keys on exact integer wei while the refine pass computes its sizes in
# floats, so a ladder node landing on a grid node survives as a near-duplicate --
# measured on tac, where the second copy read as saturation, capped the arc at a
# tenth of the trade, and reported "src not connected to dst" on a chain whose
# pool quotes the swap happily.
#
# A millionth is far below any deliberate spacing (the grid is decades apart, the
# ladder doubles) and far above the wei-level noise this exists to absorb.
DUPLICATE_TOL = 1e-6
# A local `B` that jumps by more than this between grid nodes is a liquidity
# cliff -- for stableswap, the edge of the peg (§2.5).
PEG_JUMP = 4.0


class CalibrationError(ValueError):
    pass


@dataclass(frozen=True, slots=True)
class Calibration:
    """The result of one arc's ladder.

    `B >= 0` and `clamped => isfinite(cap)` are enforced here, in the *only*
    place that produces a `B`.  That is what makes a negative conductance
    structurally impossible rather than something the solver has to guard
    against: `G = nu a / B` can then only be positive or (clamped) infinite,
    and the infinite case is bounded in G-space by the §9.7 ceiling.
    """

    a: float
    B: float
    cap: float = math.inf
    clamped: bool = False
    convex_flag: bool = False
    flag_reason: FlagReason = FlagReason.NONE
    drift: float = 0.0
    eta: float = math.nan
    split_hint: bool = False
    calib_delta: float = 0.0
    tangent_delta: float = 0.0
    note: str = ""

    def __post_init__(self) -> None:
        if not (self.a > 0):
            raise CalibrationError(f"a must be positive, got {self.a!r}")
        if self.B < 0:
            raise CalibrationError(
                f"B must be clamped to 0, got {self.B!r}; a negative curvature "
                "makes the Laplacian indefinite (§11.2)"
            )
        if self.clamped and not math.isfinite(self.cap):
            raise CalibrationError(
                "a clamped arc needs a finite cap: it has no self-limiting term, "
                "so a negative-eps cycle would give unbounded flow (§2.3 rule 2)"
            )

    @property
    def domain(self) -> float:
        """`d_bar = a/B`, where the quadratic model turns decreasing (M1)."""
        return math.inf if self.B <= 0 else self.a / self.B


def second_divided_differences(x: np.ndarray, y: np.ndarray) -> np.ndarray:
    """`f[x_k, x_k+1, x_k+2] = f''(xi)/2` for each consecutive triple.

    Mean-value form, so the `xi` are exact rather than asymptotic.  Computing
    these on *every* triple of a geometric grid is a strictly stronger detector
    than one 4-node ladder.
    """
    first = np.diff(y) / np.diff(x)
    return np.diff(first) / (x[2:] - x[:-2])


def calibrate(
    deltas,
    quotes,
    *,
    delta_bar: float | None = None,
    structural_flag: bool = False,
    drift_tol: float = DRIFT_TOL,
    cap: float | None = None,
    f_at_cap: float | None = None,
    quantum: float = 0.0,
) -> Calibration:
    """Fit one arc from its probe ladder.

    `deltas`/`quotes` are the *successful* probes in increasing size.  The
    origin `f(0) = 0` is free and prepended here.  `a` comes from the smallest
    node, which is why the grid escalates: 6% of mainnet arcs fail at
    `1e-6 * reserve` and a couple return zero, and a zero `a` would poison
    `log a` in the reference-price fit.
    """
    # The compiled fit, when it is installed.  Six-element arrays are the whole
    # reason: `np.diff`, `np.interp` and `np.concatenate` over six floats cost
    # 50 us a call and this runs 736 times a quote, so the dispatch is the bill
    # rather than the arithmetic.  `test_calibrate_differential.py` differs them.
    if _ACCEL_ON and _accel.available():
        got = _accel.calibrate_ladder(
            deltas, quotes, delta_bar=delta_bar,
            structural_flag=structural_flag, drift_tol=drift_tol, cap=cap,
            f_at_cap=f_at_cap, quantum=quantum)
        if got is not None:
            return got

    x = np.asarray(deltas, dtype=float)
    y = np.asarray(quotes, dtype=float)
    if x.ndim != 1 or x.shape != y.shape:
        raise CalibrationError("deltas and quotes must be 1-D and the same length")
    if x.size < 2:
        raise CalibrationError(f"need at least 2 probes, got {x.size}")
    if not np.all(np.diff(x) > 0):
        raise CalibrationError("deltas must be strictly increasing")
    if not np.all(y > 0):
        raise CalibrationError("all quotes must be positive; drop failed probes first")

    # Two nodes at the same size are one probe, and the wall test below cannot
    # read them as anything but saturation.  Collapse them first; see
    # `DUPLICATE_TOL`.
    if x.size > 1:
        distinct = np.ones(x.size, bool)
        distinct[1:] = np.diff(x) > x[:-1] * DUPLICATE_TOL
        if not distinct.all():
            x, y = x[distinct], y[distinct]
            if x.size < 2:
                raise CalibrationError(
                    "need at least 2 probes at distinct sizes, got 1 after "
                    "collapsing duplicates"
                )

    tangent_delta = float(x[0])
    a = float(y[0] / x[0])

    # --- capacity wall (§2.3 rule 2) -------------------------------------
    #
    # Truncate to the strictly-increasing prefix, and treat the last size that
    # still bought something more as a hard cap.  This is the clamped-arc shape
    # the spec provides for: no self-limiting term beyond the cap, so the cap
    # must be finite -- which it is, by construction here.
    saturated_at: float | None = None
    for k in range(1, x.size):
        if y[k] <= y[k - 1] * (1.0 + SATURATION_TOL):
            saturated_at = float(x[k - 1])
            x, y = x[:k], y[:k]
            break
    if saturated_at is not None and x.size == 1:
        # Everything above the smallest probe is flat: the arc is a fixed
        # payout, which the quadratic model cannot express at all beyond `a`.
        return Calibration(
            a=a, B=0.0, cap=saturated_at, clamped=True, convex_flag=True,
            flag_reason=FlagReason.CLAMPED, calib_delta=saturated_at,
            tangent_delta=tangent_delta, note="SATURATED",
        )

    if saturated_at is not None:
        cap = saturated_at if cap is None else min(cap, saturated_at)
    d_bar = float(delta_bar if delta_bar is not None else x[-1])
    f_bar = float(np.interp(d_bar, x, y))
    B = 2.0 * (a * d_bar - f_bar) / d_bar**2

    # --- what the output token's own resolution can fake ------------------
    #
    # `quotes` are integers of the output token, so every one is rounded down by
    # up to one unit.  For an 18-decimal token that is nothing; for GUSD's *two*
    # it is 0.01, and the ladder's small node carries it into `a`:
    #
    #     a  = y0 / x0                    error  quantum / x0
    #     B  = 2 (a d - f(d)) / d^2       error  2 (quantum/x0) / d + 2 quantum / d^2
    #
    # Measured on 3Crv -> GUSD, where the fitted curvature is -2.9e-8 against a
    # noise floor of 4.8e-8: the entire "increasing returns" the pool appeared to
    # show was the rounding, and clamping on it made a $10,000 trade unroutable.
    #
    # Below the floor the sign of the curvature is simply unknown.  Taking the
    # floor itself as `B` is the flattest reading the data supports, which is the
    # direction §2.3 clamps in -- optimistic, bounded, and finite, so the arc
    # keeps a conductance instead of needing a cap.
    noise = 0.0
    if quantum > 0:
        noise = 2.0 * (quantum / float(x[0])) / d_bar + 2.0 * quantum / d_bar**2
    quantised = B < 0.0 and -B <= noise
    if quantised:
        B = noise

    # Two node sets, for two different jobs.  Flag detection uses *every* probe,
    # so a convex patch anywhere on the sampled range is caught -- strictly
    # stronger than one 4-node ladder.
    xs = np.concatenate(([0.0], x))
    ys = np.concatenate(([0.0], y))
    D = second_divided_differences(xs, ys)
    numeric_flag = bool(np.any(D > 0) or (D.size > 1 and len(set(np.sign(D))) > 1))

    # DRIFT and eta use the local window `{0, d/4, d/2, d}` of §2.3: the origin
    # (exact, not a probe) plus the top three nodes, with the tiny tangent probe
    # deliberately excluded.  Divided differences estimate a derivative *at a
    # point*, so spreading the nodes over decades measures the spread rather than
    # the curve.  Four nodes are needed, so a two-point coarse pass reports
    # neither -- by design: they are diagnostics for arcs that carry flow.
    drift = 0.0
    eta = math.nan
    if x.size >= 3:
        xs_local = np.concatenate(([0.0], x[-3:]))
        ys_local = np.concatenate(([0.0], y[-3:]))
        D_local = second_divided_differences(xs_local, ys_local)
        if D_local[0] != 0:
            drift = float(D_local[1] / D_local[0] - 1.0)
        if D_local[1] != 0:
            D3 = float((D_local[1] - D_local[0]) / (xs_local[3] - xs_local[0]))
            eta = 3.0 * a * D3 / (2.0 * D_local[1] ** 2)

    note = "SATURATED" if saturated_at is not None else ""
    if quantised:
        note = "QUANTISED"
    # A wall is a cap even when the fitted curvature is healthy: without it the
    # solver sizes the arc from `B` alone and posts flow the venue cannot fill.
    clamped = B <= 0.0 or saturated_at is not None
    flag_reason = FlagReason.NONE
    if numeric_flag and structural_flag:
        flag_reason = FlagReason.BOTH
    elif numeric_flag:
        flag_reason = FlagReason.DIVIDED_DIFF
    elif structural_flag:
        flag_reason = FlagReason.STRUCTURAL

    if clamped:
        # §2.3 zero-curvature clamp.  B = 0 is the *admissible limit* -- a
        # linear arc, no price impact, an ideal diode with no series
        # resistance -- and it is exactly the local concave envelope.
        if cap is None or not math.isfinite(cap):
            # The convex region always terminates, and the default cap is the
            # last ladder node we have data for.
            cap = float(x[-1])
            f_at_cap = float(y[-1])
            note = "CAP_FROM_LADDER"
        if f_at_cap is None:
            f_at_cap = float(np.interp(cap, x, y))
        if not (f_at_cap > 0):
            raise CalibrationError("clamped arc needs a positive f(cap) for the chord")
        # CHORD, not tangent.  Clamping B while keeping a = f'(0) leaves the
        # tangent, which on a convex piece lies *below* the curve -- an
        # under-estimate that makes the solver silently skip the arc.  The chord
        # is exact at both endpoints and above the curve in between: a valid
        # concave majorant, hence an upper bound, hence it cannot prune the true
        # optimum.
        a = float(f_at_cap / cap)
        B = 0.0
        flag_reason = FlagReason.CLAMPED if flag_reason is FlagReason.NONE else FlagReason.BOTH

    return Calibration(
        a=a,
        B=max(B, 0.0),
        cap=float(cap) if cap is not None else math.inf,
        clamped=clamped,
        convex_flag=bool(numeric_flag or structural_flag or clamped),
        flag_reason=flag_reason,
        drift=drift,
        eta=eta,
        split_hint=bool(abs(drift) > drift_tol),
        calib_delta=d_bar,
        tangent_delta=tangent_delta,
        note=note,
    )


def peg_boundary(deltas, quotes, *, jump: float = PEG_JUMP) -> float | None:
    """Where local curvature jumps -- the edge of a stableswap's flat region.

    §2.5 calls a single fit across this boundary "the most dangerous single
    mis-calibration in the system": inside the peg the pool looks bottomless, and
    any trade that leaves it is wildly over-promised.  The geometric grid turns
    detecting it into a comparison of second divided differences rather than a
    hope that `d_bar` landed inside the flat region.

    Returns the delta at which to split the arc, or None.
    """
    x = np.concatenate(([0.0], np.asarray(deltas, float)))
    y = np.concatenate(([0.0], np.asarray(quotes, float)))
    if x.size < 4:
        return None
    D = np.abs(second_divided_differences(x, y))
    for k in range(1, D.size):
        if D[k - 1] > 0 and D[k] > jump * D[k - 1]:
            return float(x[k + 1])
    return None


def asym(a_f: float, a_r: float, B_f: float, B_r: float) -> float:
    """`log(B_r/B_f) + 1.5 log(a_f/a_r)` -- zero for any smooth symmetric-fee CFMM.

    Needs no reference prices, and holds regardless of curve shape, reserve
    ratio or fee level.  Nonzero means *genuine* directional asymmetry (a v3
    tick distribution, a stableswap peg side, a dynamic fee, a transfer tax,
    or integer rounding), which means `B` must be calibrated per direction and
    never derived from the other one by (M8).

    Use tangent values: secant fits anchor at different points on different
    curves, so their conductances legitimately differ.
    """
    if min(a_f, a_r, B_f, B_r) <= 0:
        return math.nan
    return math.log(B_r / B_f) + 1.5 * math.log(a_f / a_r)


def model_output(calibration: Calibration, delta: float) -> float:
    """`f_hat(delta) = a delta - B delta^2 / 2`, clipped to the model domain."""
    d = min(delta, calibration.domain)
    return calibration.a * d - 0.5 * calibration.B * d * d
